### PR TITLE
[Nox.Generator] Localization, OData Get support localized fields

### LINKED ISSUE #937: [Nox.Generator] Localization, OData Get support localized fields
- relates to #907 
- depends #936 

# API Get Endpoints

Entity data and localized data will be partitioned into two tables and all localized fields will be found in a single row on the \<Entity\>Localized table, making queries very efficient, and OData compatible, example, using Linq to SQL :

 ```
  IQueryable<WorkplaceDto> linqQuery =                    
     from workplace in _appDbContext.Workplaces.FromSql($@"
               SELECT 
               Workplaces.Id
                  ,Workplaces.Name
                  ,Workplaces.BrandId
                  ,Workplaces.MyCountryCode
                  ,Workplaces.DurationTicks
                  ,Workplaces.IsDeleted
                  ,COALESCE(WorkplacesLocalized.Description,Workplaces.Description ) as Description
                  ,COALESCE(WorkplacesLocalized.ShortDescription,Workplaces.ShortDescription ) as ShortDescription                                        
                 FROM Workplaces 
                 Left join  WorkplacesLocalized 
                      ON Workplaces.Id = WorkplacesLocalized.WorkplaceDtoId
                  AND WorkplacesLocalized.LanguageCode = {languageCode}                        
                 ")

     select workplace;

 return Ok(linqQuery);
```

## Tasks
1. Get the request language from the header, fall back to default language in Nox.Solution if not found, see https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Language
   1. Use the new template for the API, migrate this method to the new template first (Create a PR just with this)  
3. Update Get\<Entity\>Query to receive a culture code if contains a localized field 
4. Update Get\<Entity\>Query  update the IQueryable to return the proper translated fields according to the culturecode
   1. Generate SQL for all columns and localized columns, if no localized columns keep as it is no changed if needed
   2. Produce SQL must be targeting the correct SQL Provider, I tested https://sqlkata.com/ seems more easy to do it ourselves, our scenarios are simple

6. if no translation is found return default between square brackets "[country description]"
1. Ensure to have tests on the client API and Integration Tests asserting the localized data is correct

### PR DESCRIPTION
**Pull request by ILoveMyTortoise**
closes #937

### CODE REVIEW

**Comment by ILoveMyTortoise** [comment]
TODO: discuss short path with Ricardo

File: src/Nox.Types.Abstractions/TypeDefinitions/NoxSimpleTypeDefinition.cs [hunk 1]
```diff
@@ -129,6 +129,8 @@ public class NoxSimpleTypeDefinition
     [Description("Indicates whether this attribute is readonly. Defaults to false.")]
     public bool IsReadonly { get; internal set; } = false;
 
+    public bool ShouldBeLocalized => Type == NoxType.Text && TextTypeOptions!.IsLocalized;
```
Comment (rochar) [inline src/Nox.Types.Abstractions/TypeDefinitions/NoxSimpleTypeDefinition.cs]:
```suggestion
    public bool IsLocalized=> Type == NoxType.Text && TextTypeOptions!.IsLocalized;
```

File: src/Nox.Solution/Models/Domain/Entities/Entity.cs [hunk 2]
```diff
@@ -105,7 +106,7 @@ public class Entity : DefinitionBase
     public bool ShouldBeLocalized =>
```
Comment (rochar) [inline src/Nox.Solution/Models/Domain/Entities/Entity.cs]:
```suggestion
    public bool IsLocalized=>
```

File: src/Nox.ClientApi.Tests/Generated/Nox.Generator/Nox.Generator.NoxCodeGenerator/Application.Queries/GetWorkplacesQuery.g.cs [hunk 3]
```diff
@@ -8,33 +8,64 @@
...
+using Nox.Solution;
 
 namespace ClientApi.Application.Queries;
 
 public record GetWorkplacesQuery() : IRequest<IQueryable<WorkplaceDto>>;
 
 internal partial class GetWorkplacesQueryHandler: GetWorkplacesQueryHandlerBase
 {
-    public GetWorkplacesQueryHandler(DtoDbContext dataDbContext): base(dataDbContext)
+    public GetWorkplacesQueryHandler(DtoDbContext dataDbContext,
```
Comment (rochar) [inline src/Nox.ClientApi.Tests/Generated/Nox.Generator/Nox.Generator.NoxCodeGenerator/Application.Queries/GetWorkplacesQuery.g.cs]: Quesries are Application Level, they can not depend on HTTP function, they can be invoked by any component, for example a IntegrationEvent Consumer

The query needs to receive a Nox.Type.CultureCode

File: src/Nox.ClientApi.Tests/Generated/Nox.Generator/Nox.Generator.NoxCodeGenerator/Application.Queries/GetWorkplacesQuery.g.cs [hunk 4]
```diff
@@ -8,33 +8,64 @@
...
     public DtoDbContext DataDbContext { get; }
 
     public virtual Task<IQueryable<WorkplaceDto>> Handle(GetWorkplacesQuery request, CancellationToken cancellationToken)
     {
-        var item = (IQueryable<WorkplaceDto>)DataDbContext.Workplaces
-            .AsNoTracking();
-       return Task.FromResult(OnResponse(item));
+        var cultureCode = _languageProvider.GetLanguage();
+
+        IQueryable<WorkplaceDto> linqQueryBuilder =
```
Comment (rochar) [inline src/Nox.ClientApi.Tests/Generated/Nox.Generator/Nox.Generator.NoxCodeGenerator/Application.Queries/GetWorkplacesQuery.g.cs:48]: Can you add to the PR comments the SQL generated so its easy to understand the final result

File: src/Nox.ClientApi.Tests/Generated/Nox.Generator/Nox.Generator.NoxCodeGenerator/Application.Queries/GetWorkplacesQuery.g.cs [hunk 5]
```diff
@@ -8,33 +8,64 @@
...
+            from item in DataDbContext.Workplaces.AsNoTracking()
+            join itemLocalizedFromJoin in DataDbContext.WorkplacesLocalized on cultureCode equals itemLocalizedFromJoin.CultureCode into joinedData
+            from itemLocalized in joinedData.Where(l => item.Id == l.Id).DefaultIfEmpty()
```
Comment (rochar) [inline src/Nox.ClientApi.Tests/Generated/Nox.Generator/Nox.Generator.NoxCodeGenerator/Application.Queries/GetWorkplacesQuery.g.cs:51]: this produces one full join and one left join, one left join should be sufficient, can we discuss this?

File: src/Nox.ClientApi.Tests/Generated/Nox.Generator/Nox.Generator.NoxCodeGenerator/Application.Queries/GetWorkplacesQuery.g.cs [hunk 6]
```diff
@@ -8,33 +8,64 @@
...
-    public  GetWorkplacesQueryHandlerBase(DtoDbContext dataDbContext)
+{private readonly NoxSolution _solution;
+        private readonly IHttpLanguageProvider _languageProvider;
+
+    public  GetWorkplacesQueryHandlerBase(DtoDbContext dataDbContext,
+        NoxSolution solution,
+        IHttpLanguageProvider languageProvider)
     {
-        DataDbContext = dataDbContext;
+        DataDbContext = dataDbContext;_solution = solution;
```
Comment (rochar) [inline src/Nox.ClientApi.Tests/Generated/Nox.Generator/Nox.Generator.NoxCodeGenerator/Application.Queries/GetWorkplacesQuery.g.cs]: Add a new line between statements

File: src/Nox.ClientApi.Tests/Generated/Nox.Generator/Nox.Generator.NoxCodeGenerator/Application.Queries/GetWorkplacesQuery.g.cs [hunk 7]
```diff
@@ -8,33 +8,64 @@
...
 }
 
 internal abstract class GetWorkplacesQueryHandlerBase : QueryBase<IQueryable<WorkplaceDto>>, IRequestHandler<GetWorkplacesQuery, IQueryable<WorkplaceDto>>
-{
-    public  GetWorkplacesQueryHandlerBase(DtoDbContext dataDbContext)
+{private readonly NoxSolution _solution;
+        private readonly IHttpLanguageProvider _languageProvider;
+
+    public  GetWorkplacesQueryHandlerBase(DtoDbContext dataDbContext,
+        NoxSolution solution,
```
Comment (rochar) [inline src/Nox.ClientApi.Tests/Generated/Nox.Generator/Nox.Generator.NoxCodeGenerator/Application.Queries/GetWorkplacesQuery.g.cs]: NoxSolution is not being used, can we remove it?

**Review by rochar — CHANGES_REQUESTED** [review]
we are on a good track, some comment to review and discuss

File: src/Nox.ClientApi.Tests/Generated/Nox.Generator/Nox.Generator.NoxCodeGenerator/Application.Queries/GetWorkplacesQuery.g.cs [hunk 8]
```diff
@@ -8,33 +8,64 @@
...
     public virtual Task<IQueryable<WorkplaceDto>> Handle(GetWorkplacesQuery request, CancellationToken cancellationToken)
     {
-        var item = (IQueryable<WorkplaceDto>)DataDbContext.Workplaces
-            .AsNoTracking();
-       return Task.FromResult(OnResponse(item));
+        var cultureCode = _languageProvider.GetLanguage();
+
+        IQueryable<WorkplaceDto> linqQueryBuilder =
+            from item in DataDbContext.Workplaces.AsNoTracking()
+            join itemLocalizedFromJoin in DataDbContext.WorkplacesLocalized on cultureCode equals itemLocalizedFromJoin.CultureCode into joinedData
```
Comment (rochar) [inline src/Nox.ClientApi.Tests/Generated/Nox.Generator/Nox.Generator.NoxCodeGenerator/Application.Queries/GetWorkplacesQuery.g.cs:50]:
```suggestion
            join itemLocalizedFromJoin in DataDbContext.WorkplacesLocalized on cultureCode equals '###CULTURECODE###' into joinedData
```

-- @__cultureCode_0='en-US'
SELECT w."Id", w."Name", COALESCE(t."Description", ('[' || COALESCE(w."Description", '')) || ']') AS "Description", w."Greeting", w."BelongsToCountryId", w."Etag"
FROM "Workplaces" AS w
LEFT JOIN (
    SELECT w0."Id", w0."Description"
    FROM "WorkplacesLocalized" AS w0
    WHERE 'en-US' = w0."CultureCode"
) AS t ON w."Id" = t."Id"

**Comment by ILoveMyTortoise** [comment]
generated sql

-- @__cultureCode_0='en-US'
SELECT w."Id", w."Name", COALESCE(t."Description", ('[' || COALESCE(w."Description", '')) || ']') AS "Description", w."Greeting", w."BelongsToCountryId", w."Etag"
FROM "Workplaces" AS w
LEFT JOIN (
    SELECT w0."Id", w0."Description"
    FROM "WorkplacesLocalized" AS w0
    WHERE 'en-US' = w0."CultureCode"
) AS t ON w."Id" = t."Id"

File: src/Nox.ClientApi.Tests/Generated/Nox.Generator/Nox.Generator.NoxCodeGenerator/Application.Queries/GetWorkplacesQuery.g.cs (hunk 7 above)
Comment (ILoveMyTortoise) [reply src/Nox.ClientApi.Tests/Generated/Nox.Generator/Nox.Generator.NoxCodeGenerator/Application.Queries/GetWorkplacesQuery.g.cs]: Done

File: src/Nox.ClientApi.Tests/Generated/Nox.Generator/Nox.Generator.NoxCodeGenerator/Application.Queries/GetWorkplacesQuery.g.cs (hunk 6 above)
Comment (ILoveMyTortoise) [reply src/Nox.ClientApi.Tests/Generated/Nox.Generator/Nox.Generator.NoxCodeGenerator/Application.Queries/GetWorkplacesQuery.g.cs]: Done

File: src/Nox.ClientApi.Tests/Generated/Nox.Generator/Nox.Generator.NoxCodeGenerator/Application.Queries/GetWorkplacesQuery.g.cs (hunk 5 above)
Comment (ILoveMyTortoise) [reply src/Nox.ClientApi.Tests/Generated/Nox.Generator/Nox.Generator.NoxCodeGenerator/Application.Queries/GetWorkplacesQuery.g.cs:51]: Done

File: src/Nox.ClientApi.Tests/Generated/Nox.Generator/Nox.Generator.NoxCodeGenerator/Application.Queries/GetWorkplacesQuery.g.cs (hunk 4 above)
Comment (ILoveMyTortoise) [reply src/Nox.ClientApi.Tests/Generated/Nox.Generator/Nox.Generator.NoxCodeGenerator/Application.Queries/GetWorkplacesQuery.g.cs:48]: Done

File: src/Nox.ClientApi.Tests/Generated/Nox.Generator/Nox.Generator.NoxCodeGenerator/Application.Queries/GetWorkplacesQuery.g.cs (hunk 3 above)
Comment (ILoveMyTortoise) [reply src/Nox.ClientApi.Tests/Generated/Nox.Generator/Nox.Generator.NoxCodeGenerator/Application.Queries/GetWorkplacesQuery.g.cs]: Done

File: src/Nox.Solution/Models/Domain/Entities/Entity.cs (hunk 2 above)
Comment (ILoveMyTortoise) [reply src/Nox.Solution/Models/Domain/Entities/Entity.cs]: Done

File: src/Nox.Types.Abstractions/TypeDefinitions/NoxSimpleTypeDefinition.cs (hunk 1 above)
Comment (ILoveMyTortoise) [reply src/Nox.Types.Abstractions/TypeDefinitions/NoxSimpleTypeDefinition.cs]: Done

File: src/Nox.ClientApi.Tests/Generated/Nox.Generator/Nox.Generator.NoxCodeGenerator/Application.Queries/GetWorkplacesQuery.g.cs (hunk 8 above)
Comment (ILoveMyTortoise) [reply src/Nox.ClientApi.Tests/Generated/Nox.Generator/Nox.Generator.NoxCodeGenerator/Application.Queries/GetWorkplacesQuery.g.cs:50]: Let's discuss on call

**Review by rochar — APPROVED** [review]

**Pull request merged**